### DISCUSSION TITLE
Use of `save-input` and `restore-input` in programs

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#134] [General]
The standard words [`save-input`](https://forth-standard.org/standard/core/SAVE-INPUT) and [`restore-input`](https://forth-standard.org/standard/core/RESTORE-INPUT) are conceptually like [`file-position`](https://forth-standard.org/standard/file/FILE-POSITION) and [`reposition-file`](https://forth-standard.org/standard/file/REPOSITION-FILE), but they work for the current input source only, and the input source position is provided as an opaque data object that cannot be altered.

Did anybody use these words in programs or libraries? (not in system internals) 

I also asked this question in comp.lang.forth, and only one example [was mentioned](https://groups.google.com/g/comp.lang.forth/c/AqQrQ4rx6rI/m/K13vsAtTAAAJ) at the moment — "Lambda Expressions in ANS Forth" by @gerryjackson ([sources](http://qlikz.org/forth/archive/lambda.zip)), and in this example the problem can be also solved without these words.

**Comment by SirWumpus** [comment]
I've seen examples of rescanning the current `SOURCE` (input line), by push / alter / pop operations on `>IN`, which is fine for the current contents of `SOURCE`.

Having reread `SAVE-INPUT`, `RESTORE-INPUT`, and their rationale, there is no mention as to behaviour of `RESTORE-INPUT` if a 
text or block file input is repositioned earlier than the current line found in `SOURCE` or likewise read past the current line.  

In these cases the `SOURCE` contents will have been altered.  `RESTORE-INPUT` will restore the positional data, but there is no mention if it should reread (text or block file) input in order to ensure `SOURCE` contents and `>IN` reflect the state at the time of `SAVE-INPUT`.   String input is not a problem and the terminal is well understood.

If `SAVE-INPUT` /  `RESTORE-INPUT` are only useful for the current `SOURCE` contents, then they're seem pretty pointless (vs just manipulating `>IN`).

**Reply by ruv** [reply]
>  there is no mention if it should reread (text or block file) input in order to ensure `SOURCE` contents and `>IN` reflect the state at the time of `SAVE-INPUT`.

I cannot agree on that.

[`source`](https://forth-standard.org/standard/core/SOURCE) returns an address and length of _the input buffer_. I think, a state of the input buffer includes the **content** of the input buffer.  The different content means the different state. And a state of the input buffer is a part of the _input source specification_ (by its definition).

Also, have a look on the testing section, [F.6.2.2182 SAVE-INPUT](https://forth-standard.org/standard/testsuite#test:core:SAVE-INPUT). The test will fail if the content of the input buffer is not restored.

In a standard program the **content** of the input buffer can be only changed via [`refill`](https://forth-standard.org/standard/core/REFILL) (always forward repositioning), or `restore-input` (can be forward or backward repositioning, but not earlier the point where `save-input` was performed at the first time on the current input source). NB: after performing such words like `evaluate` or `included` the content of the input buffer is always the same as before performing (i.e., it's not changed).

It's not necessary that the content of the input buffer is reread from mass-storage during `restore-input`. Technically, this content can be saved in a new buffer, and then this buffer is made the input buffer on `restore-input`. The life time of this new buffer should be not less than the life time of the corresponding input source (taking into account that `restore-input` can be applied many times to the same input source specification).

**Reply by SirWumpus** [reply]
https://github.com/SirWumpus/post4/commit/c287a4d932f8505838185c416a196c02bde30867#r99633655

> My original thought was to re-read from storage, but one of your suggestions to squirrel away the current buffer without I/O overhead seemed like a nice idea. Clearly these two words are painful and why they appear in the standard if they are seldom used seems like bad idea, leading to confusion (example being me), and mental pain & suffering similar to a night at an opera. Shouldn't a case be made to deprecate these words from the [next] Draft?
>
> Starting to lean more towards `kill -9 save-input restore-input`.

**Comment by gerryjackson** [comment]
On 02/02/2023 14:50, ruv wrote:
>
>     there is no mention if it should reread (text or block file) input
>     in order to ensure |SOURCE| contents and |>IN| reflect the state
>     at the time of |SAVE-INPUT|.
>
> I cannot agree on that.
>
> |source| <https://forth-standard.org/standard/core/SOURCE> returns an 
> address and length of /the input buffer/. I think, a state of the 
> input buffer includes the *content* of the input buffer. The different 
> content means the different state. And a state of the input buffer is 
> a part of the /input source specification/ (by its definition).
>
> Also, have a look on the testing section, F.6.2.2182 SAVE-INPUT 
> <https://forth-standard.org/standard/testsuite#test:core:SAVE-INPUT>. 
> The test will fail if the content of the input buffer is not restored.
>
> In a standard program the *content* of the input buffer can be only 
> changed via |refill| <https://forth-standard.org/standard/core/REFILL> 
> (always forward repositioning), or |restore-input|. After performing 
> such words like |evaluate| or |included| the content of the input 
> buffer is always the same as before performing.
>
> It's not necessary that the content of the input buffer is reread form 
> mass-storage during |restore-input|. Technically, this content can be 
> saved in a new buffer, and then this buffer is made the input buffer 
> on |restore-input|. The life time of this new buffer should be not 
> less than the life time if the corresponding input source.
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/ForthHub/discussion/discussions/134#discussioncomment-4852455>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ADH775PJF7Q3IHRHSCB7J6DWVPCT7ANCNFSM6AAAAAAUOHFMJI>.
> You are receiving this because you are subscribed to this thread.
>
Related to this is the fact that SOURCE-ID returns /fileid/ for a file 
source. A few years ago I wanted to use the /fileid/ to reposition the 
file position to the end of the file and reported an error to a Forth 
system when it didn't work. The Forth 200X committee discussed this and 
came to the conclusion that when including a file the programmer had no 
entitlement to use the /fileid/ for REPOSITION-FILE etc. That led to the 
question why the /fileid/ was returned by SOURCE-ID if a program 
couldn't use it. That was never answered as far as I know. That 
judgement seems wrong to me.

Gerry

**Reply by SirWumpus** [reply]
I agree.  I saw that this morning, Draft 19-1 11.3.3 Input source :

> During text interpretation from a text file, the value returned by FILE-POSITION for the fileid returned by SOURCE-ID is undefined. A standard program shall not call REPOSITION-FILE on the fileid returned by SOURCE-ID.

Seems perversely wrong.

**Reply by SirWumpus** [reply]
Where else is one suppose to obtain the fileid of a text file for `SAVE-INPUT`?

**Reply by ruv** [reply]
@gerryjackson writes:
>  why the /fileid/ was returned by SOURCE-ID if a program couldn't use it.

I guess, it's just an omission. If you allow a program to apply `reposition-file` (as well as other FILE words that change the state of a file, e.g. `close-file`) to this _fileid_, you should specify how it does affect the input buffer, parse area, and `refill`.  Otherwise you will get different effects on different systems. So it's safe to apply to this _fileid_ the only words that don't change the state of a file (like `file-position` or `file-size`).

BTW, answers via GitHub (instead of email) looks slightly more pretty ([example](https://github.com/ForthHub/discussion/discussions/134#discussioncomment-4852455)).

-----

@SirWumpus writes:

> Draft 19-1 11.3.3 Input source

Yes, it was not changed from [11.3.3 in Forth-2012](https://forth-standard.org/standard/file#subsection.11.3.3), which says the same.

> Where else is one suppose to obtain the _fileid_ of a text file for `SAVE-INPUT`?

It's an  internal business of the Forth system, up to the implementer.  The only question is whether this _fileid_ should be exposed to programs or not. Now `source-id` cannot be changed to return something else instead of a _fileid_.
